CSLabs.Api/Controllers/ModuleController.cs

```diff
@@ -117,9 +133,16 @@ public async Task<IActionResult> Upsert([FromBody] Module module)
                 DatabaseContext.Update(module);
             else
                 DatabaseContext.Add(module);
+            // add tag relationships
+            await DatabaseContext.AddRangeAsync(module.ModuleTags);
```

Comment:
You are calling add on a list of module tags that may or may not need to be inserted.
The following call might have been enough.
```
DatabaseContext.Update(module);
```

In order to get an accurate representation of what would happen, we need to test the method somehow. You could test it using postman but that could be unwieldy since you will need to write the request json by hand. You can also test it by going ahead and building the frontend piece for editing tags in the module editor. It will be much easier for you to figure out what you need to happen. You could also build an integration test to debug the method.


Once you have  a way to make requests, you will want to take a look under the hood and figure out what entity framework is going to do when  `SaveChangesAsync` is called by looking at the `ChangeTracker`.

Start a debugging session in rider, and place a breakpoint on `DatabaseContext.Update(module);` and `await DatabaseContext.AddRangeAsync(module.ModuleTags);`. Make a request using the module editor on the frontend and view the change tracker state when the breakpoints are hit.
Use the evaluation window button in rider and paste the following to show the tracked changed as a list.
```
DatabaseContext.ChangeTracker.Entries().ToList()
```

![image](https://user-images.githubusercontent.com/4037224/127685116-fa59ea58-e1b5-4f83-b746-341a24eb7a5c.png)


if you can't figure out the evaulation window, you can assign it to a variable and see the variable's value in th debugger.
```
var entriesBefore = DatabaseContext.ChangeTracker.Entries().ToList()
```

Update might try to insert them all regardless if they are already in the database and result in error.
You will probably need to loop through them and insert if their id is equal to 0 else update.